### PR TITLE
Add tests

### LINKED ISSUE #6: Use pytest as testing framework
In #2, the "tests" are a file to be run with Python. [pytest](https://docs.pytest.org/en/latest/) provides a lot of nice features like running *all* test files, specifying which tests exactly are failing, and much more. To do so, tests should be placed in files called `test_something.py`, and within each file, each test should be its own function, `def test_something_specific():`, which can have assertions etc within it.

### PR DESCRIPTION
**Pull request by tlambert03**
closes #6  
starting work on tests with pytest.  To try them out, pull this PR, do `pip install pytest pytest-benchmark`, then run `python -m pytest`.  (after there is a setup.py file in here, you can install the package locally with `pip install -e .` and then just "pytest" alone will work).

I added an example benchmark to compare CPU with GPU at different array sizes for test_multiply_matrix ... which do look like there's some work to be done  😅
```
---------------------------------------- benchmark: 16 tests --------------------------
Name (time in us)                              Mean                 StdDev       Rounds
---------------------------------------------------------------------------------------
test_multiply_matrix[cpu-64x64]             70.9399 (1.0)           7.6045         6413
test_multiply_matrix[gpu-64x64]          3,693.2913 (52.06)       259.3121          230
test_multiply_matrix[cpu-128x128]          267.7748 (3.77)         26.8374         2153
test_multiply_matrix[gpu-128x128]        3,987.1152 (56.20)       270.0672          240
test_multiply_matrix[cpu-256x256]        1,139.2580 (16.06)       180.4972          619
test_multiply_matrix[gpu-256x256]        5,276.0616 (74.37)       412.6583          164
test_multiply_matrix[cpu-512x512]        4,861.8999 (68.54)       601.1902          158
test_multiply_matrix[gpu-512x512]       11,101.6707 (156.49)      235.0929           76
test_multiply_matrix[cpu-1024x1024]     21,199.7104 (298.84)    1,290.4645           30
test_multiply_matrix[gpu-1024x1024]     44,066.2162 (621.18)    1,059.5423           20
test_multiply_matrix[cpu-2048x2048]    126,025.3478 (>1000.0)   8,929.1151            6
test_multiply_matrix[gpu-2048x2048]    210,418.4442 (>1000.0)  11,606.9894            5
test_multiply_matrix[cpu-4096x4096]    814,364.7442 (>1000.0)  42,047.9409            5
test_multiply_matrix[gpu-4096x4096]  1,161,997.9468 (>1000.0)  16,589.3983            5
test_multiply_matrix[cpu-8192x8192]  4,916,867.3818 (>1000.0)  56,979.3099            5
test_multiply_matrix[gpu-8192x8192]  6,693,529.5528 (>1000.0)  65,086.0641            5
----------------------------------------------------------------------------------------
```

`test_kernels` still needs a lot of work, and feel free to directly edit this PR if you're so inclined.

### CODE REVIEW

**Comment by haesleinhuepf** [comment]
Hey @tlambert03 ,

> starting work on tests with pytest.

awesome, thanks! I started [cleaning up](https://github.com/clEsperanto/pyclesperanto_prototype/commit/09f93abb2377195e96dd20bb6b070e7d1c22cec2) this huge single test file already. These things will integrate nicely.

> I added an example benchmark to compare CPU with GPU at different array sizes for test_multiply_matrix ... which do look like there's some work to be done 😅

Hehe, I also did some [benchmarks](https://github.com/clEsperanto/pyclesperanto_prototype/tree/tlambert-forks-setup.py/benchmarks) - which were highly suspicious: My Intel GPU appears to be faster than the NVidia. I assume under the hood of pyopencl/gputools some config is wrongly set. But we'll figure that out.

So that's a beautiful example benchmark with the matrix-multiplication. You got me 😉  The [open-cl code](https://github.com/clEsperanto/clij-opencl-kernels/blob/master/src/main/java/net/haesleinhuepf/clij/kernels/multiply_matrix_x.cl) for this one is zero optimized. There is an [opportunity](https://cnugteren.github.io/tutorial/pages/page2.html) waiting ;-) 

Matrix-multiplication was never a bottleneck in any of my workflows, hence I never spent time in optimizing it. I realized quite early during CLIJ development: If I optimize all kernels, it will take me 48 years. Thus, I concentrated on the operations in my workflows which were really slow and a bottleneck to our research. You know, we mostly do image segmentation / quantitative analysis on 3D light sheet data. Thus, I was much more worried (I was never _worried_) about [connected components](https://github.com/clij/clij2/blob/master/src/main/java/net/haesleinhuepf/clij2/plugins/ConnectedComponentsLabeling.java#L47) and [pixel statistics](https://github.com/clij/clij2/blob/master/src/main/java/net/haesleinhuepf/clij2/plugins/StatisticsOfLabelledPixels.java#L386) (this one you know as regionprops) because these are the operations you can't typically simply parallelize on a GPU. Not talking about [fill holes](https://github.com/clij/clij2-docs/blob/master/src/main/macro/benchmarkFillHoles.ijm), [voronoi](https://github.com/clij/clij2-docs/blob/master/src/main/macro/benchmarkVoronoi.ijm) and [skeletonize](https://github.com/clij/clij2-docs/blob/master/src/main/macro/benchmarkSkeletonize.ijm) 😓 .

Btw. know your hardware ;-) This is your test running on a laptop with a Core i5-8250U versus a RTX 2080 Ti eGPU😁 :

```
----------------------------------------------- benchmark: 16 tests -------------------------------------
Name (time in us)                                    Mean                  StdDev                  Rounds
---------------------------------------------------------------------------------------------------------
test_multiply_matrix[cpu-64x64]                   74.5375 (1.0)            7.5263 (1.0)                32
test_multiply_matrix[cpu-128x128]                342.4807 (4.59)          96.5161 (12.82)             622
test_multiply_matrix[cpu-256x256]              1,466.8504 (19.68)        371.9322 (49.42)             456
test_multiply_matrix[gpu-128x128]              6,787.6054 (91.06)        333.2698 (44.28)              37
test_multiply_matrix[gpu-64x64]                6,921.5615 (92.86)        139.3366 (18.51)              13
test_multiply_matrix[gpu-256x256]              7,779.4892 (104.37)       272.9641 (36.27)              37
test_multiply_matrix[cpu-512x512]              9,611.1875 (128.94)     1,852.7705 (246.17)            128
test_multiply_matrix[gpu-512x512]             20,909.3500 (280.52)       336.2859 (44.68)              26
test_multiply_matrix[cpu-1024x1024]           41,649.8143 (558.78)     3,722.6038 (494.61)             28
test_multiply_matrix[gpu-1024x1024]           66,899.2200 (897.52)     7,724.1440 (>1000.0)             5
test_multiply_matrix[cpu-2048x2048]          203,129.8000 (>1000.0)   12,337.9847 (>1000.0)             6
test_multiply_matrix[gpu-2048x2048]          295,492.7800 (>1000.0)   18,606.2453 (>1000.0)             5
test_multiply_matrix[cpu-4096x4096]        1,093,684.3600 (>1000.0)   47,568.2510 (>1000.0)             5
test_multiply_matrix[gpu-4096x4096]        1,625,478.3400 (>1000.0)   64,907.9132 (>1000.0)             5
test_multiply_matrix[gpu-8192x8192]        9,185,164.5800 (>1000.0)   98,980.6824 (>1000.0)             5
test_multiply_matrix[cpu-8192x8192]       11,202,774.5200 (>1000.0)  385,148.4120 (>1000.0)             5
---------------------------------------------------------------------------------------------------------
```

Btw this goes pretty much in line with some earlier tests I did [in matlab](https://github.com/clij/clatlab/blob/master/src/main/matlab/matrix_multiply_benchmarking.m): Starting at matrix size of `(n*1000) x (n*1000)` (`5 < n < 10` depends on hardware) the naive implementation on the GPU starts outperforming the heavily engineered solution in matlab and numpy. I'm actually quite proud of this 🙂  The effort-gain-ratio is on my side 🤣

File: tests/test_add_images.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+input1 = np.asarray([[1, 2, 3]])
+input2 = np.asarray([[4, 5, 6]])
+
+
+def test_add_images_weighted_missing_params():
+    reference = np.asarray([[5, 7, 9]])
+    output = cle.add_images_weighted(input1, input2)
+    result = cle.pull(output)
+    assert np.array_equal(result, reference)
```
Comment (lagru) [inline tests/test_add_images.py:14]: You might want to use [np.testing.assert_equal](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_equal.html#numpy.testing.assert_equal) here and below instead?

Comment (tlambert03) [reply tests/test_add_images.py:14]: agreed.  I started changing everything to the `np.testing` module in `tests/test_cle.py` ... and so far was just "pytestifying" the code in `test_add_images` before cleaning it up

**Comment by tlambert03** [comment]
> My Intel GPU appears to be faster than the NVidia. I assume under the hood of pyopencl/gputools some config is wrongly set. But we'll figure that out.

This is exactly what I'm suggesting.  I've seen much better improvements on the same hardware in my own projects using pyopencl directly.  So was just marking this as something to look into.  probably a combination of the kernel itself and the helper logic.

> Btw. know your hardware

do you suspect that I don't 😅 ?  I've been coding for pyopencl/pycuda for quite a while now and am acutely aware of the hardware dependency (I wasn't just testing on my macbook hoping the integrated graphics on the CPU would do magic).  This test was on a GTX 1080 Ti (the best one I have that still works with hackintosh... which I am running), certainly not a RTX 2080, but no slouch either.  I can test with windows/RTX Titan as well later

**Comment by haesleinhuepf** [comment]
> This test was on a i7 6700k vs a GTX 1080 Ti (the best one I have that still works with hackintosh... which I am running)

I have had the suspicion that it picks the Intel GPU instead of the RTX...

**Comment by tlambert03** [comment]
ah, that may be improved by the `get_best_device` function that I added in #13 ... but I'll double check to see what the default device is in the current repo vs that function there

**Comment by tlambert03** [comment]
on my system it's picking the better GPU:  if I log the device used in the test i get:
```
-------- available devices -----------
platform: 	Apple
device type: 	CPU
	Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz 
device type: 	GPU
	GeForce GTX 1080 Ti 

-------- currently used device -------
NAME: 	  GeForce GTX 1080 Ti
GLOBAL_MEM_SIZE: 	  11811160064
GLOBAL_MEM_SIZE: 	  11811160064
MAX_MEM_ALLOC_SIZE: 	  2952790016
LOCAL_MEM_SIZE: 	  49152
IMAGE2D_MAX_WIDTH: 	  32768
IMAGE2D_MAX_HEIGHT: 	  32768
IMAGE3D_MAX_WIDTH: 	  16384
IMAGE3D_MAX_HEIGHT: 	  16384
IMAGE3D_MAX_DEPTH: 	  2048
MAX_WORK_GROUP_SIZE: 	  1024
MAX_WORK_ITEM_SIZES: 	  [1024, 1024, 64]
```

File: tests/test_add_images.py (hunk 1 above)
Comment (haesleinhuepf) [reply tests/test_add_images.py:14]: May I ask why? Curious

**Comment by haesleinhuepf** [comment]
> if I log the device used

How do you do this? I always wanted to know :-)

File: tests/test_add_images.py (hunk 1 above)
Comment (tlambert03) [reply tests/test_add_images.py:14]: it's honestly not _that_ big of a difference, and there are pros/cons and differing opinions on each side.  See, for example, this thread: https://github.com/scipy/scipy/issues/8583
and discussion here: https://numpy.org/doc/stable/reference/testing.html
I have a slight preference for np.testing ... for point # 2 in that first link above:

> Guaranteed support. numpy.testing can be guaranteed to work as long as numpy does, whereas assert relies on pytest maintenance and/or pytest-compatible assert support by any future testing framework.

**Comment by tlambert03** [comment]
well, there's the quick/dirty `print` statement, or the python [`logging`](https://docs.python.org/3/library/logging.html) module, which looks like this:
```
import logging
logger = logging.getLogger(__name__)  # general pattern, where __name__ gives the current module name

#then to use it
logger.debug("only shows if I turn up the logging level")
logger.info("just some info")
logger.warn("something might be wrong")
# etc...
```

as for _where_ you put this,  I put it in `OCLProgram` right after `get_device()`.  Since this is still using gputools directly, you either need to have a local editable (pip install -e) version of gputools installed, _or_ (and it might be slightly surprising that you can do this) you can directly edit the pip-installed gputools source code.  If you're using an IDE, you can usually right click on `OCLProgram` and go to definition, which will bring you to the installed package ... or, you can directly find the `site-packages` folder in your conda environment (can't remember the direct path on windows at the moment), find the gputools package, and start adding statements.

This is another argument for getting that `OCLProgram` class within this package though :)

**Comment by haesleinhuepf** [comment]
> This is another argument for getting that `OCLProgram` class within this package though :)

Yes. And this device logging thing must beome to `cle.clInfo();` I'll see if I manage that... ;-) Thanks so far. So good to have some people around who know how this all is supposed to work :-)

**Comment by haesleinhuepf** [comment]
`cle.cl_info();` pardon :-)

**Comment by tlambert03** [comment]
also forgot to add, if you start sprinkling code with various logging levels and you want to see the debug statements, you need to change the global logging level with something like [`logging.basicConfig(level='DEBUG')`](https://docs.python.org/3/library/logging.html#logging.basicConfig), or, with pytest, you can change `--log-level` on the command line

**Comment by haesleinhuepf** [comment]
Hey @tlambert03 

I might need your help: I'm trying to merge all this together and I'm struggling with simple stuff: I'm now on another branch trying to execute [this](https://github.com/clEsperanto/pyclesperanto_prototype/blob/tlambert-forks-setup.py/playground.py) with:

```
python playground.py
```

And receive this error message:

```
Traceback (most recent call last):
  File "playground.py", line 1, in <module>
    import pyclesperanto_prototype as cle
  File "C:\structure\code\pyclesperanto_prototype\pyclesperanto_prototype\__init__.py", line 1, in <module>
    from .tier0 import *
  File "C:\structure\code\pyclesperanto_prototype\pyclesperanto_prototype\tier0\__init__.py", line 3, in <module>
    from .create import create
  File "C:\structure\code\pyclesperanto_prototype\pyclesperanto_prototype\tier0\create.py", line 1, in <module>
    from ._pycl import OCLArray
  File "C:\structure\code\pyclesperanto_prototype\pyclesperanto_prototype\tier0\_pycl.py", line 126, in <module>
    if cl.characterize.has_double_support(get_gpu().device):
AttributeError: module 'pyopencl' has no attribute 'characterize'
```

I do have this pyopencl installed:

```
pyopencl                  2020.2+cl12              pypi_0    pypi
```

May this come from the gputools-versus-_pycl [change](https://github.com/clEsperanto/pyclesperanto_prototype/commit/b084f0b4da03c1ac138727bd7861f14c2fb1d5d2)? 😕

**Comment by tlambert03** [comment]
oh sorry, I think that was my bad... looks like `pyopencl.__init__` doesn't export the characterize module:
in the `_pycl.py` file, put `from pyopencl import characterize` at the top,
and then in line 126 change 
```
if cl.characterize.has_double_support(get_gpu().device):
``` 
to 
```
if characterize.has_double_support(get_gpu().device):
```

**Comment by tlambert03** [comment]
> `cle.cl_info();` pardon :-)

here's a start:

```python
import pyopencl as cl


def device_info(dev: cl.Device):
    info = dict()
    for c in dir(cl.device_info):
        if c[0].isupper() and c != "PARENT_DEVICE":
            try:
                info[c] = dev.get_info(getattr(cl.device_info, c))
            except cl.LogicError:
                info[c] = None
    return info


def platform_info(platform: cl.Platform):
    info = dict()
    for c in dir(cl.platform_info):
        if c[0].isupper():
            try:
                info[c] = platform.get_info(getattr(cl.platform_info, c))
            except cl.LogicError:
                info[c] = None
    return info


def cl_info():
    lines = []
    for platform in cl.get_platforms():
        lines += [platform.name]
        lines += [f"{k}:{v}" for k, v in platform_info(platform).items()]
        lines += ["\n"]
        for device in platform.get_devices():
            lines += ["   " + device.name]
            lines += [f"   {k}:{v}" for k, v in device_info(device).items()]
            lines += ["\n"]
        lines += ["\n"]
    print("\n".join(lines))

```

**Comment by haesleinhuepf** [comment]
closing this because changes are also in #19

**Pull request closed without merging**